lib/graphql/analysis/ast.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+require "graphql/analysis/ast/visitor"
+require "graphql/analysis/ast/analyzer"
+require "graphql/analysis/ast/field_usage"
+require "graphql/analysis/ast/query_complexity"
+require "graphql/analysis/ast/max_query_complexity"
+require "graphql/analysis/ast/query_depth"
+require "graphql/analysis/ast/max_query_depth"
+
+# frozen_string_literal: true
+module GraphQL
+  module Analysis
+    module AST
+      module_function
+
+      # @return [void]
+      def analyze_multiplex(multiplex, analyzers)
+        multiplex_analyzers = analyzers.map { |analyzer| analyzer.new(multiplex) }
+
+        multiplex.trace("analyze_multiplex", { multiplex: multiplex }) do
+          query_results = multiplex.queries.map do |query|
+            if query.valid?
+              analyze_query(
+                query,
+                query.ast_analyzers,
+                multiplex_analyzers: multiplex_analyzers
+              )
+            else
+              []
+            end
+          end
+
+          multiplex_results = multiplex_analyzers.map(&:result)
+          multiplex_errors = analysis_errors(multiplex_results)
+
+          multiplex.queries.each_with_index do |query, idx|
+            query.analysis_errors = multiplex_errors + analysis_errors(query_results[idx])
+          end
+        end
+        nil
+      end
+
+      # @param query [GraphQL::Query]
+      # @param analyzers [Array<#call>] Objects that respond to `#call(memo, visit_type, irep_node)`
+      # @return [Array<Any>] Results from those analyzers
+      def analyze_query(query, analyzers, multiplex_analyzers: [])
+        query.trace("analyze_query", { query: query }) do
```

Comment:
Could we trace each individual analyzer as well?

Reply:
The thing is, they're called all together, during one visit. They don't make  sequential visits. I did that originally because I thought it would be faster, but if I'm wrong, or if it's not worth the tradeoff, I'm open to reconsidering the design.

Reply:
Oh I think that's fine then. If its all a single visit, then the overhead within a single analyzer would likely be more obvious (like calling a service or something which should be done after the query anyway?).